Client/WarFare/UIItemUpgrade.cpp

```diff
@@ -0,0 +1,1333 @@
+			&& m_pSlotArea[i]->IsIn(ptCur.x, ptCur.y))
+			return UIWND_DISTRICT_UPGRADE_SLOT;
+	}
+
+	for (int i = 0; i < MAX_ITEM_INVENTORY; i++)
+	{
+		if (m_pInvArea[i] != nullptr
+			&& m_pInvArea[i]->IsIn(ptCur.x, ptCur.y))
+			return UIWND_DISTRICT_UPGRADE_INV;
+	}
+
+	return UIWND_DISTRICT_UNKNOWN;
+}
+
+// Handles UI messages such as button clicks and icon events.
+bool CUIItemUpgrade::ReceiveMessage(CN3UIBase* pSender, uint32_t dwMsg)
+{
+	if (pSender == nullptr)
+		return false;
+
+	if (dwMsg == UIMSG_BUTTON_CLICK)
+	{
+		if (pSender == m_pBtnClose)
+		{
+			Close();
+		}
+		else if (pSender == m_pBtnCancel)
+		{
+			if (!m_bUpgradeInProgress)
+				ResetUpgradeInventory();
+		}
+		else if (pSender == m_pBtnOk)
+		{
+			if (!m_bUpgradeInProgress)
+				SendToServerUpgradeMsg();
+		}
+
+		return true;
+	}
+
+	__IconItemSkill* spItem = nullptr;
+	e_UIWND_DISTRICT  eUIWnd = UIWND_DISTRICT_UNKNOWN;
+
+	int iOrder = -1;
+
+	switch (dwMsg)
+	{
+		case UIMSG_ICON_DOWN_FIRST:
+		{
+			spItem = m_pSelectedItem;
+			iOrder = m_iSelectedItemSourcePos;
+			eUIWnd = GetWndDistrict();
+
+			if (eUIWnd == UIWND_DISTRICT_UPGRADE_RESULT_SLOT)
+				ResetUpgradeInventory();
+
+			if (iOrder == -1
+				|| eUIWnd != UIWND_DISTRICT_UPGRADE_INV)
+			{
+				SetState(UI_STATE_COMMON_NONE);
+				return false;
+			}
+
+			// Divide countable items
+			if (spItem->iCount > 1
+				&& spItem->IsStackable())
+			{
+				__IconItemSkill* pNew = new __IconItemSkill(*spItem);
+				CreateUIIconForItem(pNew);
+				ShowItemCount(spItem, iOrder);
+				m_pSelectedItem = pNew;
+			}
+
+			// Set icon region for moving.
+			RECT region = GetSampleRect();
+			m_pSelectedItem->pUIIcon->SetRegion(region);
+			m_pSelectedItem->pUIIcon->SetMoveRect(region);
+		}
+		break;
+
+		case UIMSG_ICON_UP:
+			spItem = m_pSelectedItem;
+			iOrder = m_iSelectedItemSourcePos;
+			if (spItem == nullptr)
+				break;
+
+			if (!ReceiveIconDrop(spItem))
+			{
+				if (spItem->iCount > 1
+					&& spItem->IsStackable())
+				{
+					ShowItemCount(spItem, iOrder);
+
+					// Clean divided item
+					if (spItem != nullptr)
+					{
+						if (spItem->pUIIcon != nullptr)
+						{
+							delete spItem->pUIIcon;
+							spItem->pUIIcon = nullptr;
+						}
+
+						delete spItem;
+						spItem = nullptr;
+					}
+				}
+				else
+				{
+					// Restore the icon position to its original place if drop failed.
+					CancelIconDrop(spItem);
+				}
+			}
+			break;
+
+		case UIMSG_ICON_DOWN:
+			spItem = m_pSelectedItem;
+
+			if (GetState() == UI_STATE_ICON_MOVING
+				&& spItem != nullptr
+				&& spItem->pUIIcon != nullptr)
+			{
+				RECT region = GetSampleRect();
+				spItem->pUIIcon->SetRegion(region);
+				spItem->pUIIcon->SetMoveRect(region);
+			}
+			break;
+
+		case UIMSG_ICON_RDOWN_FIRST:
+			if (!m_bUpgradeInProgress)
+				HandleInventoryIconRightClick(m_pSelectedItem);
+			break;
+	}
+	return true;
+}
+
+void CUIItemUpgrade::SetVisibleWithNoSound(bool bVisible, bool bWork, bool bReFocus)
```

Comment:
This should be sure to focus (`CGameProcedure::s_pUIMgr->SetVisibleFocusedUI(this);`) when `bReFocus` is true, and reset focus on hide (when `bVisible` is false and `bReFocus` is true).